### PR TITLE
Fix/consistentbasewidget

### PR DESCRIPTION
**Pull request by andydotxyz**
### Description:
Move our internal Base widget (that could not be extended) to be consistent with the public BaseWidget.
Now we have just 1 API and the ability to extend both.
It is still 2 different classes due to the cycle issues. One cannot be an alias of the other due to certain unexported APIs.
This will be further tidied before 2.1 by hopefully replacing the `ExtendBaseWidget` idea - but for now it should unblock some keyboard operation items.

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### CODE REVIEW

File: internal/widget/base.go [hunk 1]
```diff
@@ -8,115 +8,153 @@ import (
...
 
-// Base is a simple base widget for internal use.
+// Base provides a helper that handles basic widget behaviours.
 type Base struct {
-	hidden bool
-	pos    fyne.Position
-	size   fyne.Size
+	size     fyne.Size
+	position fyne.Position
+	hidden   bool
```
Comment (toaster) [inline internal/widget/base.go:15]: FMI: why did you remove the alphabetical order?

File: internal/widget/scroller.go [hunk 2]
```diff
@@ -538,16 +442,13 @@ func (s *Scroll) Refresh() {
 	s.refreshWithoutOffsetUpdate()
 }
 
-func (s *Scroll) refreshWithoutOffsetUpdate() {
-	RefreshWidget(s)
+func (s *Scroll) Resize(sz fyne.Size) {
+	s.Base.Resize(sz)
+	s.Refresh()
```
Comment (toaster) [inline internal/widget/scroller.go:448]: Shouldn’t we keep the refreshes limited to actual changes of the size (`size != s.size`)?

File: internal/driver/glfw/menu_bar.go [hunk 3]
```diff
@@ -199,6 +160,7 @@ func (r *menuBarRenderer) Refresh() {
 	r.Layout(r.b.Size())
 	r.background.FillColor = theme.ButtonColor()
 	r.background.Refresh()
+	r.ShadowingRenderer.RefreshShadow()
```
Comment (toaster) [inline internal/driver/glfw/menu_bar.go:163]: FYI: I guess, this is not really related to the base widget change and therefore should have been in a separate commit, right?

File: widget/tree.go [hunk 4]
```diff
@@ -643,6 +643,12 @@ func (n *treeNode) Tapped(*fyne.PointEvent) {
 	n.tree.Select(n.uid)
 }
 
+func (n *treeNode) Resize(s fyne.Size) {
+	n.BaseWidget.Resize(s)
+	// TODO find out if the indentation changed. If not, and size is the same, then we can skip this as normal
```
Comment (toaster) [inline widget/tree.go:648]: TBH, I think that a caller of `Resize` should never rely on the refresh. If the indentation gets changed at some point and depends on the refresh of a `Resize` call with the same size, this is wrong. Instead the change of the indentation should trigger a refresh.

File: internal/driver/glfw/menu_bar_item.go [hunk 5]
```diff
@@ -113,6 +113,11 @@ func (i *menuBarItem) MouseOut() {
 // Implements: fyne.Tappable
 func (i *menuBarItem) Tapped(*fyne.PointEvent) {
 	i.Parent.toggle(i)
+
+	if i.Parent.active {
+		i.Parent.canvas.Focus(i)
+	}
+	i.Refresh()
```
Comment (toaster) [inline internal/driver/glfw/menu_bar_item.go:120]: There is no test for this, therefore I don’t know which bug you’re trying to fix with it.
But I think this is partly wrong. The parent’s `toggle` already focuses the item on activation and unfocuses it on deactivation.
So, if the menu is active after the toggle, the item should already be focused and the added focus code does nothing.
If the additional `Refresh` fixes some bug, we should try to add a test case for it and also move it to `MenuBar`’s `toggle()`. Currently, the item gets refreshed before it is being focused. Maybe that’s to early?

**Review by toaster — CHANGES_REQUESTED** [review]
A couple of questions/suggestions as normal comments (not needed for my approval).
And one objection to the menu bar refresh fix.

File: internal/driver/glfw/menu_bar.go (hunk 3 above)
Comment (andydotxyz) [reply internal/driver/glfw/menu_bar.go:164]: I had assumed it was something that I broke in the refactoring. If it was already broken then yes it should be a separate commit.
I don't think it's worth unpicking and force pushing a better version, but I will try to keep an eye out in future commits.
If you disagree say so and I'll rework the history

File: internal/driver/glfw/menu_bar_item.go (hunk 5 above)
Comment (andydotxyz) [reply internal/driver/glfw/menu_bar_item.go]: Apologies it was not a bug fix, I had misread the code in the refactor and you are right it should not have been needed. What has changed is that on tapped the focus in is not triggering so we don't get the same refresh request cycle. I tracked down that `activateChild` does not actually refresh the button - only the bar. The bar's refresh does not cascade to the list of items it would seem. I have put in place what feels like a better fix in `activateChild`, what do you think?

File: internal/widget/base.go (hunk 1 above)
Comment (andydotxyz) [reply internal/widget/base.go]: A mistake, sorry

File: internal/widget/scroller.go (hunk 2 above)
Comment (andydotxyz) [reply internal/widget/scroller.go:452]: Yes, good catch thanks.

File: widget/tree.go (hunk 4 above)
Comment (andydotxyz) [reply widget/tree.go:648]: It's not relying on a Refresh - it's the Layout that we care about.
Normally we can ignore Resize if the size has not changed - but in the case of Tree it may be the same size but need to have the content moved when it's reused (because increased indent does not change size).
Maybe there is a better solution, but I cannot think of one right now.

**Review by andydotxyz — COMMENTED** [review]
Thanks for the notes, I hope these comments and the commit help :)

File: widget/tree.go (hunk 4 above)
Comment (toaster) [reply widget/tree.go:648]: I still don’t fully understand why a `Resize` of a tree node has to handle indentation change. AFAICS an indentation change is implemented by the node’s `update` method which should then do the appropriate things.
Also, AFAIK, the `BaseWidget`'s `Resize` doesn’t do anything if the size didn’t change (apart from heating the world).

Comment (andydotxyz) [reply widget/tree.go]: Quite right, I managed to track down the root cause. The `leaf` and `branch` widgets were having the `treeNode` widget inside accessed directly, which was part of the issue. The rest was that the `partialRefresh` called when depth changed was not updating the node layout.

With this in place it should be fixed much more satisfactorily.

**Review by Jacalz — APPROVED** [review]
This looks good and sensible to me, though I would have preferred to get rid of `.ExtendBaseWidget()` entirely ;)

**Comment by andydotxyz** [comment]
> This looks good and sensible to me, though I would have preferred to get rid of `.ExtendBaseWidget()` entirely ;)

Yes, that is the plan. But it will be easier to get rid of from a consistent place :).
Plus I don't yet know how we will do it, but I do know that this was blocking.

**Comment by toaster** [comment]
> This looks good and sensible to me, though I would have preferred to get rid of `.ExtendBaseWidget()` entirely ;)

Yeah, that’s something @andydotxyz and me discussed already. Sadly, it’s not trivial.

**Review by toaster — APPROVED** [review]
